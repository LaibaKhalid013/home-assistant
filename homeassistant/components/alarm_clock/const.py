"""Constants for the Alarm Clock integration."""

import voluptuous as vol

from homeassistant.const import CONF_ICON, CONF_NAME
import homeassistant.helpers.config_validation as cv

DOMAIN = "alarm_clock"
ENTITY_ID_FORMAT = DOMAIN + ".{}"

ATTR_ALARM_TIME = "alarm_time"
ATTR_REPEAT_DAYS = "repeat_days"
ATTR_NEXT_ALARM = "next_alarm"

CONF_ALARM_TIME = "alarm_time"
CONF_REPEAT_DAYS = "repeat_days"

EVENT_ALARM_CLOCK_STARTED = "alarm_clock.started"
EVENT_ALARM_CLOCK_FINISHED = "alarm_clock.finished"
EVENT_ALARM_CLOCK_CANCELLED = "alarm_clock.cancelled"
EVENT_ALARM_CLOCK_CHANGED = "alarm_clock.changed"

STORAGE_KEY = DOMAIN
STORAGE_VERSION = 1

STORAGE_FIELDS = {
    vol.Required(CONF_NAME): cv.string,
    vol.Optional(CONF_ICON): cv.icon,
    vol.Required(CONF_ALARM_TIME): cv.time,
    vol.Required(CONF_REPEAT_DAYS): cv.weekdays,
}
