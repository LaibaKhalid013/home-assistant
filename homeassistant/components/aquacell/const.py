"""Constants for the Aquacell integration."""

DOMAIN = "aquacell"
DATA_AQUACELL = "DATA_AQUACELL"
