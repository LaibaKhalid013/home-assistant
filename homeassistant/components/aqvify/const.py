"""Constants for the Aqvify integration."""

import logging

DOMAIN = "aqvify"

CONF_DEVICE_KEY = "device_key"

LOGGER = logging.getLogger(__package__)
