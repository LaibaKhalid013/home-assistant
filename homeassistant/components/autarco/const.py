"""Constants for the Autarco integration."""

from __future__ import annotations

from datetime import timedelta
import logging
from typing import Final

DOMAIN: Final = "autarco"
LOGGER = logging.getLogger(__package__)
SCAN_INTERVAL = timedelta(minutes=1)

CONF_PUBLIC_KEY: Final = "public_key"

SENSORS_SOLAR: Final = "solar"
