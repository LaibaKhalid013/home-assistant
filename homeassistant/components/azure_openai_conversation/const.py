"""Constants for the OpenAI Conversation integration."""

import logging

DOMAIN = "azure_openai_conversation"
LOGGER = logging.getLogger(__package__)

CONF_AZURE_OPENAI_RESOURCE = "azure_openai_resource"
CONF_RECOMMENDED = "recommended"
CONF_PROMPT = "prompt"
CONF_CHAT_MODEL = "chat_model"
RECOMMENDED_CHAT_MODEL = "gpt-4"
CONF_MAX_TOKENS = "max_tokens"
RECOMMENDED_MAX_TOKENS = 150
CONF_TOP_P = "top_p"
RECOMMENDED_TOP_P = 1.0
CONF_TEMPERATURE = "temperature"
RECOMMENDED_TEMPERATURE = 1.0
AZURE_OPEN_API_VERSION = "2024-02-01"
