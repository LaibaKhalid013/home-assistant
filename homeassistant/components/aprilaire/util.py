"""Utililities for the Aprilaire integration."""

from math import ceil, floor

from homeassistant.const import UnitOfTemperature
from homeassistant.util.unit_conversion import TemperatureConverter


def convert_temperature_if_needed(
    temperature_unit: UnitOfTemperature, temperature: float | None
) -> float | None:
    """Convert a temperature manually to correct rounding errors.

    This is due to the fact that the rounding in Home Assistant does not match the rounding
    that is performed on the device. Relying on the Home Assistant rounding results in an
    off-by-one issue. This code uses the current temperature unit for the sensor and rounds
    according to the Aprilaire conventions to match what is displayed on the device.
    """

    if temperature is not None and temperature_unit == UnitOfTemperature.FAHRENHEIT:
        raw_fahrenheit = TemperatureConverter.convert(
            temperature, UnitOfTemperature.CELSIUS, UnitOfTemperature.FAHRENHEIT
        )

        if raw_fahrenheit >= 0:
            temperature = floor(raw_fahrenheit + 0.5)
        else:
            temperature = ceil(raw_fahrenheit - 0.5)

    return temperature
