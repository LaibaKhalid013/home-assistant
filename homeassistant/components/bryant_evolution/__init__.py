"""The Bryant Evolution integration."""

from __future__ import annotations

import logging

from evolutionhttp import BryantEvolutionClient

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST, Platform
from homeassistant.core import HomeAssistant

from .const import CONF_SYSTEM_ID, CONF_ZONE_ID

PLATFORMS: list[Platform] = [Platform.CLIMATE]

type BryantEvolutionConfigEntry = ConfigEntry[BryantEvolutionClient]  # noqa: F821
_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant, entry: BryantEvolutionConfigEntry
) -> bool:
    """Set up Bryant Evolution from a config entry."""
    entry.runtime_data = BryantEvolutionClient(entry.data[CONF_HOST], entry.data[CONF_SYSTEM_ID], entry.data[CONF_ZONE_ID])
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(
    hass: HomeAssistant, entry: BryantEvolutionConfigEntry
) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
