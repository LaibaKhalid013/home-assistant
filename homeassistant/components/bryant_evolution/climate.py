"""Support for Bryant Evolution HVAC systems."""

from datetime import timedelta
import logging
from typing import Any

from evolutionhttp import BryantEvolutionClient

from homeassistant.components.climate import (
    ClimateEntity,
    ClimateEntityFeature,
    HVACAction,
    HVACMode,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST, UnitOfTemperature
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import CONF_SYSTEM_ID, CONF_ZONE_ID

_LOGGER = logging.getLogger(__name__)

type BryantEvolutionConfigEntry = ConfigEntry[BryantEvolutionClient]  # noqa: F821
SCAN_INTERVAL = timedelta(seconds=60)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: BryantEvolutionConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up a config entry."""
    host = config_entry.data[CONF_HOST]
    system_id = config_entry.data[CONF_SYSTEM_ID]
    zone_id = config_entry.data[CONF_ZONE_ID]
    client = config_entry.runtime_data
    climate = BryantEvolutionClimate(host, system_id, zone_id, client)
    async_add_entities([climate], update_before_add=True)


class BryantEvolutionClimate(ClimateEntity):
    """ClimateEntity for Bryant Evolution HVAC systems."""

    def __init__(
        self: ClimateEntity,
        host: str,
        system_id: int,
        zone_id: int,
        client: BryantEvolutionClient,
    ) -> None:
        """Initialize an entity from parts."""
        self._host = host
        self._system_id = system_id
        self._zone_id = zone_id
        self._client = client
        self._name = f"Bryant Evolution (System {system_id}, Zone {zone_id})"
        self._attr_current_temperature = None
        self._attr_target_temperature = None
        self._attr_target_temperature_low = None
        self._attr_target_temperature_high = None
        self._attr_hvac_mode = None
        self._attr_fan_mode = None
        self._attr_hvac_action = None
        self._attr_temperature_unit = UnitOfTemperature.FAHRENHEIT
        self._attr_hvac_action = None
        self._attr_supported_features = (
            ClimateEntityFeature.TARGET_TEMPERATURE
            | ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
            | ClimateEntityFeature.FAN_MODE
            | ClimateEntityFeature.TURN_ON
            | ClimateEntityFeature.TURN_OFF
        )
        self._enable_turn_on_off_backwards_compatibility = False

    @property
    def name(self) -> str:
        """Return the name of the entity."""
        return self._name

    async def async_update(self) -> None:
        """Update the entity state."""
        self._attr_current_temperature = await self._client.read_current_temperature()
        self._attr_fan_mode = await self._client.read_fan_mode()

        self._attr_target_temperature = None
        self._attr_target_temperature_high = None
        self._attr_target_temperature_low = None
        self._attr_hvac_mode = await self._read_hvac_mode()

        # Set target_temperature or target_temperature_{high, low} based on mode.
        match self._attr_hvac_mode:
            case HVACMode.HEAT:
                self._attr_target_temperature = (
                    await self._client.read_heating_setpoint()
                )
            case HVACMode.COOL:
                self._attr_target_temperature = (
                    await self._client.read_cooling_setpoint()
                )
            case HVACMode.HEAT_COOL:
                self._attr_target_temperature_high = (
                    await self._client.read_cooling_setpoint()
                )
                self._attr_target_temperature_low = (
                    await self._client.read_heating_setpoint()
                )
            case HVACMode.OFF:
                pass
            case _:
                _LOGGER.error("Unknown HVAC mode %s", self._attr_hvac_mode)

        # Note: depends on current temperature and target temperature low read
        # above.
        self._attr_hvac_action = await self._read_hvac_action()
        _LOGGER.error("XXX async update done")

    @property
    def unique_id(self) -> str:
        """Return a unique ID for this entity."""
        return f"bryant_evolution_{self._host}_{self._system_id}_{self._zone_id}"

    async def _read_hvac_mode(self) -> HVACMode:
        mode_and_active = await self._client.read_hvac_mode()
        if not mode_and_active:
            raise HomeAssistantError("Failed to read current HVAC mode")
        mode = mode_and_active[0]

        # Response format is MODE and (optionally) a number indicating which
        # system is currently calling for the MODE (e.g., HEAT 1 if system 1
        # is calling for heat; HEAT if no call is being made).
        match mode.upper():
            case "HEAT":
                return HVACMode.HEAT
            case "COOL":
                return HVACMode.COOL
            case "AUTO":
                return HVACMode.HEAT_COOL
            case "OFF":
                return HVACMode.OFF

        raise HomeAssistantError(f"Cannot parse response to HVACMode: {mode}")

    async def _read_hvac_action(self) -> HVACAction:
        """Return the current running hvac operation."""
        mode_and_active = await self._client.read_hvac_mode()
        if not mode_and_active:
            raise HomeAssistantError("Failed to read current HVAC action")
        mode, is_active = mode_and_active
        if not is_active:
            return HVACAction.OFF
        match mode.upper():
            case "HEAT":
                return HVACAction.HEATING
            case "COOL":
                return HVACAction.COOLING
            case "OFF":
                return HVACAction.OFF
            case "AUTO":
                # In AUTO, we need to figure out what the actual mode is
                # based on the setpoints.
                if (
                    self.current_temperature is not None
                    and self.target_temperature_low is not None
                ):
                    if self.current_temperature > self.target_temperature_low:
                        # If the system is on and the current temperature is
                        # higher than the point at which heating would activate,
                        # then we must be cooling.
                        return HVACAction.COOLING
                    return HVACAction.HEATING
        raise HomeAssistantError(
            f"Could not determine HVAC action: {mode_and_active}, {self.current_temperature}, {self.target_temperature_low}"
        )

    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
        """Set new target hvac mode."""
        if hvac_mode == HVACMode.HEAT_COOL:
            hvac_mode = HVACMode.AUTO
        if not await self._client.set_hvac_mode(hvac_mode):
            raise HomeAssistantError("Failed to set HVAC mode")
        self._attr_hvac_mode = hvac_mode
        self._async_write_ha_state()

    async def async_set_temperature(self, **kwargs: Any) -> None:
        """Set new target temperature."""
        if kwargs.get("target_temp_high"):
            temp = int(kwargs["target_temp_high"])
            if not await self._client.set_cooling_setpoint(temp):
                raise HomeAssistantError("Failed to set cooling setpoint")
            self._attr_target_temperature_high = temp
            self._async_write_ha_state()

        if kwargs.get("target_temp_low"):
            temp = int(kwargs["target_temp_low"])
            if not await self._client.set_heating_setpoint(temp):
                raise HomeAssistantError("Failed to set heating setpoint")
            self._attr_target_temperature_low = temp
            self._async_write_ha_state()

        if kwargs.get("temperature"):
            temp = int(kwargs["temperature"])
            fn = (
                self._client.set_heating_setpoint
                if self.hvac_mode == HVACMode.HEAT
                else self._client.set_cooling_setpoint
            )
            if not await fn(temp):
                raise HomeAssistantError("Failed to set temperature")
            self._attr_target_temperature = temp
            self._async_write_ha_state()

    async def async_set_fan_mode(self, fan_mode: str) -> None:
        """Set new target fan mode."""
        if not await self._client.set_fan_mode(fan_mode):
            raise HomeAssistantError("Failed to set fan mode")
        self._attr_fan_mode = fan_mode
        self.async_write_ha_state()

    @property
    def hvac_modes(self) -> list[HVACMode]:
        """Return the list of available HVAC operation modes."""
        return [HVACMode.HEAT, HVACMode.COOL, HVACMode.HEAT_COOL, HVACMode.OFF]

    @property
    def fan_modes(self) -> list[str]:
        """Return the list of available fan modes."""
        return ["AUTO", "LOW", "MED", "HIGH"]
