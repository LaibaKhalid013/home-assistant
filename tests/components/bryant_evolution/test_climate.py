"""Test the BryantEvolutionClient type."""

import asyncio
from collections.abc import Callable
from dataclasses import dataclass
import logging
from unittest.mock import patch

from evolutionhttp import BryantEvolutionLocalClient
import pytest

from homeassistant.components.bryant_evolution.const import (
    CONF_SYSTEM_ID,
    CONF_ZONE_ID,
    DOMAIN,
)
from homeassistant.components.climate import (
    ATTR_FAN_MODE,
    ATTR_HVAC_ACTION,
    SERVICE_SET_FAN_MODE,
    SERVICE_SET_HVAC_MODE,
    SERVICE_SET_TEMPERATURE,
    HVACAction,
    HVACMode,
)
from homeassistant.config_entries import ConfigEntryState
from homeassistant.const import ATTR_ENTITY_ID, CONF_FILENAME
from homeassistant.core import HomeAssistant
from homeassistant.util.unit_system import US_CUSTOMARY_SYSTEM

from tests.common import MockConfigEntry

_LOGGER = logging.getLogger(__name__)


async def _wait_for_cond(predicate: Callable[[], bool]) -> None:
    """Wait for `predicate` to return True."""
    while True:
        if predicate():
            break
        await asyncio.sleep(0.1)


class _FakeEvolutionClient(BryantEvolutionLocalClient):
    """Fake version of `BryantEvolutionClient.

    Important design point: this type provides a `set_allow_reads` method.
    When False, all read_* calls will hang. This allows testing that service
    call handlers on the BryantEvolutionClient type properly call
    async_write_ha_state (since async_update cannot complete when reads are
    paused).
    """

    def __init__(self) -> None:
        super().__init__("UNUSED", 1, 1)
        self._temp = 75
        self._mode = "COOL"
        self._clsp = 72
        self._htsp = None
        self._fan = "AUTO"
        self._allow_reads = True
        self._allow_reads_cond = asyncio.Condition()
        self._is_active = False

    async def read_current_temperature(self) -> int | None:
        async with self._allow_reads_cond:
            await self._allow_reads_cond.wait_for(self._are_reads_allowed)
            return self._temp

    async def read_cooling_setpoint(self) -> int | None:
        async with self._allow_reads_cond:
            await self._allow_reads_cond.wait_for(self._are_reads_allowed)
            return self._clsp

    async def read_heating_setpoint(self) -> int | None:
        async with self._allow_reads_cond:
            await self._allow_reads_cond.wait_for(self._are_reads_allowed)
            return self._htsp

    def _are_reads_allowed(self) -> bool:
        return self._allow_reads

    async def read_fan_mode(self) -> str | None:
        async with self._allow_reads_cond:
            await self._allow_reads_cond.wait_for(self._are_reads_allowed)
            return self._fan

    async def set_cooling_setpoint(self, temperature: int) -> bool:
        async with self._allow_reads_cond:
            self._clsp = temperature
            return True

    async def set_heating_setpoint(self, temperature: int) -> bool:
        async with self._allow_reads_cond:
            self._htsp = temperature
            return True

    async def read_hvac_mode(self) -> tuple[str, bool] | None:
        async with self._allow_reads_cond:
            await self._allow_reads_cond.wait_for(self._are_reads_allowed)
            if not self._mode:
                return None
            return (self._mode, self._is_active)

    async def set_fan_mode(self, fan_mode: str) -> bool:
        async with self._allow_reads_cond:
            self._fan = fan_mode
            return True

    async def set_allow_reads(self, allow: bool) -> None:
        async with self._allow_reads_cond:
            self._allow_reads = allow
            self._allow_reads_cond.notify_all()

    async def set_hvac_mode(self, hvac_mode: HVACMode) -> bool:
        async with self._allow_reads_cond:
            self._mode = hvac_mode
            return True

    async def set_is_active(self, is_active: bool) -> None:
        async with self._allow_reads_cond:
            self._is_active = is_active


@pytest.fixture
async def mock_evolution_entry(
    hass: HomeAssistant,
) -> MockConfigEntry:
    """Configure and return a Bryant evolution integration."""
    hass.config.units = US_CUSTOMARY_SYSTEM
    with patch(
        "evolutionhttp.BryantEvolutionLocalClient.get_client",
        return_value=_FakeEvolutionClient(),
    ):
        entry = MockConfigEntry(
            domain=DOMAIN,
            data={CONF_FILENAME: "/dev/ttyUSB0", CONF_SYSTEM_ID: 1, CONF_ZONE_ID: 1},
        )
        entry.add_to_hass(hass)
        await hass.config_entries.async_setup(entry.entry_id)
        await hass.async_block_till_done()
        return entry


async def test_setup_integration(
    hass: HomeAssistant, mock_evolution_entry: MockConfigEntry
) -> None:
    """Test that an instance can be constructed."""
    state = hass.states.get("climate.bryant_evolution_system_1_zone_1")
    assert state, hass.states.async_all()
    assert state.state == "cool"
    assert state.attributes["fan_mode"] == "auto"
    assert state.attributes["current_temperature"] == 75
    assert state.attributes["temperature"] == 72


async def test_setup_integration_prevented_by_unavailable_client(
    hass: HomeAssistant,
) -> None:
    """Test that setup throws ConfigEntryNotReady when the client is unavailable."""
    hass.config.units = US_CUSTOMARY_SYSTEM
    mock_evolution_entry = MockConfigEntry(
        domain=DOMAIN,
        data={
            # This file does not exist.
            CONF_FILENAME: "test_setup_integration_prevented_by_unavailable_client",
            CONF_SYSTEM_ID: 1,
            CONF_ZONE_ID: 1,
        },
    )
    mock_evolution_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(mock_evolution_entry.entry_id)
    await hass.async_block_till_done()
    assert mock_evolution_entry.state is ConfigEntryState.SETUP_RETRY


async def test_setup_integration_client_returns_none(hass: HomeAssistant) -> None:
    """Test that an instance can be constructed from an unavailable client."""
    hass.config.units = US_CUSTOMARY_SYSTEM
    with patch(
        "evolutionhttp.BryantEvolutionLocalClient.get_client",
        return_value=_FakeEvolutionClient(),
    ) as p:
        mock_evolution_entry = MockConfigEntry(
            domain=DOMAIN,
            data={CONF_FILENAME: "/dev/ttyUSB0", CONF_SYSTEM_ID: 1, CONF_ZONE_ID: 1},
        )
        client = p.return_value
        client._fan = None
        client._clsp = None
        client._htsp = None
        mock_evolution_entry.add_to_hass(hass)
        await hass.config_entries.async_setup(mock_evolution_entry.entry_id)
        await hass.async_block_till_done()
        state = hass.states.get("climate.bryant_evolution_system_1_zone_1")
        assert state, hass.states.async_all()


async def test_set_temperature(
    hass: HomeAssistant, mock_evolution_entry: MockConfigEntry
) -> None:
    """Test that setting target temperature HEAT and COOL modes works."""

    @dataclass
    class TestCase:
        mode: HVACMode
        initial_temp: int
        new_temp: int
        temp_reader: Callable[[BryantEvolutionLocalClient], int]

    testcases = [
        TestCase(
            HVACMode.COOL,
            90,
            80,
            lambda client: client.read_cooling_setpoint(),
        ),
        TestCase(HVACMode.HEAT, 40, 50, lambda client: client.read_heating_setpoint()),
    ]
    client = mock_evolution_entry.runtime_data
    for case in testcases:
        # Enter case.mode with a known setpoint.
        data = {"hvac_mode": case.mode}
        data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
        await hass.services.async_call(
            "climate", SERVICE_SET_HVAC_MODE, data, blocking=True
        )
        data = {"temperature": case.initial_temp}
        data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
        await hass.services.async_call(
            "climate", SERVICE_SET_TEMPERATURE, data, blocking=True
        )
        state = hass.states.get("climate.bryant_evolution_system_1_zone_1")
        assert state.attributes["temperature"] == case.initial_temp
        assert await case.temp_reader(client) == case.initial_temp
        assert await client.read_hvac_mode() == (case.mode.lower(), False)

        # Change the setpoint, pausing reads to the device so that we
        # verify that changes are locally committed.
        await mock_evolution_entry.runtime_data.set_allow_reads(False)
        data = {"temperature": case.new_temp}
        data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
        await hass.services.async_call(
            "climate", SERVICE_SET_TEMPERATURE, data, blocking=False
        )
        state = hass.states.get("climate.bryant_evolution_system_1_zone_1")
        await _wait_for_cond(
            lambda case=case,
            s=hass.states.get("climate.bryant_evolution_system_1_zone_1"): s.attributes[
                "temperature"
            ]
            == case.new_temp
        )
        await mock_evolution_entry.runtime_data.set_allow_reads(True)
        assert await case.temp_reader(client) == case.new_temp
        await hass.async_block_till_done()


async def test_set_temperature_mode_heat_cool(
    hass: HomeAssistant, mock_evolution_entry: MockConfigEntry
) -> None:
    """Test that changing the setpoint in HEAT_COOL mode works."""
    client = mock_evolution_entry.runtime_data

    data = {"hvac_mode": HVACMode.HEAT_COOL}
    data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
    await hass.services.async_call(
        "climate", SERVICE_SET_HVAC_MODE, data, blocking=True
    )
    data = {"target_temp_low": 50, "target_temp_high": 90}
    data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
    await hass.services.async_call(
        "climate", SERVICE_SET_TEMPERATURE, data, blocking=True
    )
    state = hass.states.get("climate.bryant_evolution_system_1_zone_1")
    assert state.attributes["target_temp_low"] == 50
    assert state.attributes["target_temp_high"] == 90
    assert await client.read_cooling_setpoint() == 90
    assert await client.read_heating_setpoint() == 50
    assert await client.read_hvac_mode() == ("auto", False)

    # Change the setpoint, pausing reads to the device so that we
    # verify that changes are locally committed.
    await mock_evolution_entry.runtime_data.set_allow_reads(False)
    data = {"target_temp_low": 70, "target_temp_high": 80}
    data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
    await hass.services.async_call(
        "climate", SERVICE_SET_TEMPERATURE, data, blocking=False
    )
    await _wait_for_cond(
        lambda s=hass.states.get(
            "climate.bryant_evolution_system_1_zone_1"
        ): s.attributes["target_temp_low"] == 70
        and s.attributes["target_temp_high"] == 80
    )
    await mock_evolution_entry.runtime_data.set_allow_reads(True)
    await hass.async_block_till_done()


async def test_set_fan_mode(
    hass: HomeAssistant, mock_evolution_entry: MockConfigEntry
) -> None:
    """Test that setting fan mode works."""
    fan_modes = ["auto", "low", "med", "high"]
    client = mock_evolution_entry.runtime_data
    for mode in fan_modes:
        # Change the fan mode, pausing reads to the device so that we
        # verify that changes are locally committed.
        await mock_evolution_entry.runtime_data.set_allow_reads(False)
        data = {ATTR_FAN_MODE: mode}
        data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
        await hass.services.async_call(
            "climate", SERVICE_SET_FAN_MODE, data, blocking=False
        )
        await _wait_for_cond(
            lambda mode=mode,
            s=hass.states.get("climate.bryant_evolution_system_1_zone_1"): s.attributes[
                ATTR_FAN_MODE
            ]
            == mode
        )
        await mock_evolution_entry.runtime_data.set_allow_reads(True)
        await hass.async_block_till_done()
        assert await client.read_fan_mode() == mode


async def test_hvac_action(
    hass: HomeAssistant, mock_evolution_entry: MockConfigEntry
) -> None:
    """Test that we can read the current HVAC action."""
    client = mock_evolution_entry.runtime_data

    # Initial state should be no action.
    assert (
        hass.states.get("climate.bryant_evolution_system_1_zone_1").attributes[
            ATTR_HVAC_ACTION
        ]
        == HVACAction.OFF
    )

    # Turn on the system and verify we see an action; we change the fan
    # mode to trigger the integration to re-read the device.
    await client.set_is_active(True)
    assert await client.read_hvac_mode() == ("COOL", True)
    data = {ATTR_FAN_MODE: "auto"}
    data[ATTR_ENTITY_ID] = "climate.bryant_evolution_system_1_zone_1"
    await hass.services.async_call(
        "climate", SERVICE_SET_FAN_MODE, data, blocking=False
    )
    await _wait_for_cond(
        lambda s=hass.states.get(
            "climate.bryant_evolution_system_1_zone_1"
        ): s.attributes[ATTR_HVAC_ACTION] == HVACAction.COOLING
    )
