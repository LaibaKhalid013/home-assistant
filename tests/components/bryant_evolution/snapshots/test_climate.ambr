# serializer version: 1
# name: test_setup_integration_success
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'current_temperature': 75,
      'fan_mode': 'auto',
      'fan_modes': list([
        'auto',
        'low',
        'med',
        'high',
      ]),
      'friendly_name': 'System 1 Zone 1',
      'hvac_action': <HVACAction.OFF: 'off'>,
      'hvac_modes': list([
        <HVACMode.HEAT: 'heat'>,
        <HVACMode.COOL: 'cool'>,
        <HVACMode.HEAT_COOL: 'heat_cool'>,
        <HVACMode.OFF: 'off'>,
      ]),
      'max_temp': 95,
      'min_temp': 45,
      'supported_features': <ClimateEntityFeature: 395>,
      'target_temp_high': None,
      'target_temp_low': None,
      'temperature': 72,
    }),
    'context': <ANY>,
    'entity_id': 'climate.system_1_zone_1',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'cool',
  })
# ---
